subsys/suit/ipuc/src/suit_ipuc_sdfw.c

```diff
@@ -0,0 +1,694 @@
+		return SUIT_PLAT_ERR_UNSUPPORTED;
+	}
+
+	*slot_address = decoded_slot_address;
+	*slot_size = decoded_slot_size;
+
+	return SUIT_PLAT_SUCCESS;
+}
+
+suit_plat_err_t suit_ipuc_sdfw_get_count(size_t *count)
+{
+
+	if (count == NULL) {
+		return SUIT_PLAT_ERR_INVAL;
+	}
+
+	size_t declared_count = 0;
+
+	k_mutex_lock(&ipuc_mutex, K_FOREVER);
+	for (size_t i = 0; i < CONFIG_SUIT_IPUC_SIZE; i++) {
+		suit_ipuc_entry_t *ipuc_entry = &components[i];
+
+		if (ipuc_entry->component_id.value != NULL && ipuc_entry->component_id.len != 0) {
+			declared_count++;
+		}
+	}
+	*count = declared_count;
+	k_mutex_unlock(&ipuc_mutex);
+	return SUIT_PLAT_SUCCESS;
+}
+
+suit_plat_err_t suit_ipuc_sdfw_get_info(size_t idx, struct zcbor_string *component_id,
+					suit_manifest_role_t *role)
+{
+	if (component_id == NULL || role == NULL) {
+		return SUIT_PLAT_ERR_INVAL;
+	}
+
+	size_t cur_idx = 0;
+
+	k_mutex_lock(&ipuc_mutex, K_FOREVER);
+	for (size_t i = 0; i < CONFIG_SUIT_IPUC_SIZE; i++) {
+		suit_ipuc_entry_t *ipuc_entry = &components[i];
+
+		if (ipuc_entry->component_id.value != NULL && ipuc_entry->component_id.len != 0) {
+			if (cur_idx == idx) {
+				component_id->value = ipuc_entry->component_id.value;
+				component_id->len = ipuc_entry->component_id.len;
+				*role = ipuc_entry->role;
+				k_mutex_unlock(&ipuc_mutex);
+				return SUIT_PLAT_SUCCESS;
+			}
+			cur_idx++;
+		}
+	}
+
+	k_mutex_unlock(&ipuc_mutex);
+	return SUIT_PLAT_ERR_NOT_FOUND;
+}
+
+suit_plat_err_t suit_ipuc_sdfw_write_setup(int ipc_client_id, struct zcbor_string *component_id,
+					   struct zcbor_string *encryption_info,
+					   struct zcbor_string *compression_info)
+{
+	intptr_t slot_address = 0;
+	size_t slot_size = 0;
+
+	if (component_id == NULL) {
+		return SUIT_PLAT_ERR_INVAL;
+	}
+
+	if (ipuc_decode_mem_component_id(component_id, &slot_address, &slot_size) !=
+	    SUIT_PLAT_SUCCESS) {
+		/* Only MRAM-based MEM components are supported
+		 */
+		return SUIT_PLAT_ERR_UNSUPPORTED;
+	}
+
+	if (encryption_info != NULL && encryption_info->value != NULL &&
+	    encryption_info->len != 0) {
+		/* Encryption not supported yet
+		 */
+		return SUIT_PLAT_ERR_UNSUPPORTED;
+	}
+
+	if (compression_info != NULL && compression_info->value != NULL &&
+	    compression_info->len != 0) {
+		/* Compression not supported yet
+		 */
+		return SUIT_PLAT_ERR_UNSUPPORTED;
+	}
+
+	k_mutex_lock(&ipuc_mutex, K_FOREVER);
+	suit_ipuc_entry_t *ipuc_entry = ipuc_entry_from_component_id(component_id);
+
+	if (ipuc_entry == NULL) {
+		k_mutex_unlock(&ipuc_mutex);
+		return SUIT_PLAT_ERR_NOT_FOUND;
+	}
+
+	if (!(ipuc_entry->usage == IPUC_UNUSED || (ipuc_entry->usage == IPUC_IPC_IN_PLACE_UPDATE &&
+						   ipuc_entry->ipc_client_id == ipc_client_id))) {
+		LOG_ERR("suit_ipuc_sdfw_write_setup, IPUC address: %p, busy", (void *)slot_address);
+		k_mutex_unlock(&ipuc_mutex);
+		return SUIT_PLAT_ERR_INCORRECT_STATE;
+	}
+
+	/* Erasing if necessary
+	 */
+#define READ_CHUNK_MAX_SIZE 16
```

Comment:
Strange place to put a `#define`.
Maybe we could move the check of emptiness into a separate function, so both the define, RAM sink and the new function will be enclosed somewhere at the top of the file?